remove cases where cabal2nix produces a stringy license

jopejoe1 opened the pull request:
I'm working on removing all stringy licenses in nixpkgs, and currently the only ones left are the ones in haskellPackages ~and node2nix packages~. This change fixes the ones for Haskell packages.

sternenseemann commented on cabal2nix/src/Distribution/Nixpkgs/Haskell/FromCabal/License.hs:
```diff
@@ -40,128 +41,33 @@ fromCabalLicense AllRightsReserved                         = Known "lib.licenses
 fromCabalLicense (Apache (Just (versionNumbers -> [2,0]))) = Known "lib.licenses.asl20"
 fromCabalLicense ISC                                       = Known "lib.licenses.isc"
-fromCabalLicense OtherLicense                              = Unknown Nothing
+fromCabalLicense OtherLicense                              = Known "lib.licenses.unfree"
```
This is incorrect. We assume all licenses are not unfree since Hackage requires open source licenses.

sternenseemann commented on cabal2nix/src/Distribution/Nixpkgs/Haskell/FromCabal/License.hs:
```diff
@@ -9,24 +9,25 @@ module Distribution.Nixpkgs.Haskell.FromCabal.License
 
 fromCabalLicense :: Distribution.License.License -> Distribution.Nixpkgs.License.License
-fromCabalLicense (GPL Nothing)                             = Unknown (Just "GPL")
+fromCabalLicense (GPL Nothing)                             = Known "lib.licenses.gpl2Plus"
```
These cases are intentional as there is not enough information to conclude what variant of the license in question is used.

jopejoe1 replied:
I do realize they were intentional, but they can't stay in their current form. I'm currently working towards disallowing stringy licenses in nixpkgs, and those are pretty much the only ones left. Other options would be to make it even less informational with `lib.licenses.free` or to add a new custom license (not really a fan of this).

sternenseemann commented on cabal2nix/src/Distribution/Nixpkgs/Haskell/FromCabal/License.hs:
```diff
@@ -40,128 +41,33 @@ fromCabalLicense AllRightsReserved                         = Known "lib.licenses
-      -- Not handled: compound expressions, not expressible in Nixpkgs.
-      -- Use the SPDX expression as a free-form license string.
-      Unknown (Just $ prettyShow expr)
+fromSPDXLicense SPDX.NONE = Known "lib.licenses.unfree"
```
Again, we simply conclude this has no specified license, but is open source.

jopejoe1 replied:
Same as above, will switch this to "lib.licenses.free"

sternenseemann commented on cabal2nix/src/Distribution/Nixpkgs/Haskell/FromCabal/License.hs:
```diff
@@ -40,128 +41,33 @@ fromCabalLicense AllRightsReserved                         = Known "lib.licenses
+  _ ->
+    -- Not handed: the '+' suffix and user-defined licences references.
+    -- Use the SPDX expression as a free-form license string.
+    Unknown (Just $ DPretty.prettyShow simpl ++ " WITH " ++ DPretty.prettyShow excep)
```
Why are we reconstructing the entire expression here? We can just `prettyShow` the entire spdx expr.

jopejoe1 replied:
From what I can see at this point in the function, I do not have access to the full spdx expr, but maybe I'm missing something. This was my first time touching Haskell, still quite inexperienced with the language.

sternenseemann reviewed (changes requested):
Please also add a changelog entry, ideally noting when nixpkgs added support for `lib.licensesSpdx`.

wolfgangwalther commented on cabal2nix/src/Distribution/Nixpkgs/Haskell/FromCabal/License.hs:
```diff
@@ -40,128 +41,33 @@ fromCabalLicense AllRightsReserved                         = Known "lib.licenses
+  case simpl of
+  SPDX.ELicenseId lid -> Known ("lib.licensesSpdx.\"" ++ DPretty.prettyShow lid ++ "\" lib.licensesSpdx.\"" ++  DPretty.prettyShow excep ++ "\"")
+  _ ->
+    -- Not handed: the '+' suffix and user-defined licences references.
```
```suggestion
    -- Not handled: the '+' suffix and user-defined licences references.
    -- Use the SPDX expression as a free-form license string.
    Unknown (Just $ DPretty.prettyShow simpl)
fromSPDXLicenseExpression (SPDX.ELicense simpl (Just excep)) =
  case simpl of
  SPDX.ELicenseId lid -> Known ("lib.licensesSpdx.\"" ++ DPretty.prettyShow lid ++ "\" lib.licensesSpdx.\"" ++  DPretty.prettyShow excep ++ "\"")
  _ ->
    -- Not handled: the '+' suffix and user-defined licences references.
```

sternenseemann replied:
There is also a typo in the commit message.

wolfgangwalther commented on cabal2nix/test/golden-test-cases/highlighting-kate.nix.golden:
```diff
@@ -18,5 +18,5 @@ mkDerivation {
   homepage = "http://github.com/jgm/highlighting-kate";
   description = "Syntax highlighting";
-  license = "GPL";
+  license = lib.licenses.gpl2Plus;
```
Upstream still uses `GPL`, and has `gpl2Only`, I think: https://github.com/jgm/highlighting-kate/blob/master/LICENSE

wolfgangwalther commented on cabal2nix/test/golden-test-cases/HTF.nix.golden:
```diff
@@ -27,6 +27,6 @@ mkDerivation {
   homepage = "https://github.com/skogsbaer/HTF/";
   description = "The Haskell Test Framework";
-  license = "LGPL";
+  license = lib.licenses.lgpl2Plus;
```
Upstream has now `LGPL-2.1`, which renders as `LGPL-2.1-only` on hackage.

jopejoe1 commented:
Sorry for not being really active in here was working on an improvement for licenses in nixpkgs for making this nicer to implement here. see https://github.com/NixOS/nixpkgs/pull/468378

jopejoe1 committed:
fromSPDXLicense: handle compound licenses

make use of the new compaund licenses in nixpkgs to get more correct licensing information

jopejoe1 committed:
fromCabalLicense: return non specific varints as generic free

if we don't know the specific version of a license fallback to a generic free license

jopejoe1 committed:
fromCabalLicense: don't set any license as fallback

jopejoe1 committed:
fromCabal: don't set hydra platfrom depending on license

we do our own logic in nixpkgs based on is writen in meta.licenses with more complex logic than this

jopejoe1 committed:
cabal2nix: add info about license changes to changelog

jopejoe1 commented:
Did a larger refactor based on the changes done in https://github.com/NixOS/nixpkgs/pull/468378 and the feedback that was given here.

jopejoe1 commented on cabal2nix/src/Distribution/Nixpkgs/Haskell/FromCabal/License.hs:
```diff
@@ -40,128 +41,33 @@ fromCabalLicense AllRightsReserved                         = Known "lib.licenses
+    Unknown (Just $ DPretty.prettyShow simpl)
+fromSPDXLicenseExpression (SPDX.ELicense simpl (Just excep)) =
+  case simpl of
+  SPDX.ELicenseId lid -> Known ("lib.licensesSpdx.\"" ++ DPretty.prettyShow lid ++ "\" lib.licensesSpdx.\"" ++  DPretty.prettyShow excep ++ "\"")
```
This now always returns a valid SPDX expression; this was not possible before, as we are not allowed to have nested lists in `meta.license`

jopejoe1 committed:
fromCabalLicense: use compound license

jopejoe1 commented:
Now that compound licenses are supported in all supported nixpkgs versions, could this be moved forward?

sternenseemann commented:
Yes, I'd be interested in that!

Though, I'm still of the position that we should not fall back to something nondescript like `licenses.free` if we don't recognize the license. That can be solved manually in Nixpkgs or upstream (ideally). Probably a lot of packages affected by this would also wind up [excluded](https://github.com/NixOS/cabal2nix/pull/667) eventually.

jopejoe1 commented:
Should work like this now.
